DirectProgramming/DPC++FPGA/Tutorials/Features/prototype/CSR_and_streaming_interfaces/src/CSR_and_streaming_interfaces.cpp

```diff
@@ -0,0 +1,175 @@
+    std::string name = device.get_info<sycl::info::device::name>();
+    if (name.find(target_name) != std::string::npos) {
+      // The returned value represents a priority, this number is chosen to be
+      // large to ensure high priority
+      return 10000;
+    }
+    return -1;
+  }
+
+private:
+  std::string target_name;
+};
+
+using ValueT = int;
+// forward declare the test functions
+template <typename KernelType>
+void KernelTest(sycl::queue &, ValueT *, ValueT *, size_t);
+
+// offloaded computation
+ValueT SomethingComplicated(ValueT val) { return (ValueT)(val * (val + 1)); }
+
+/////////////////////////////////////////
+
+struct StreamingControlIP {
+  // Use the 'conduit' annotation on a kernel argument to specify it to be
+  // a streaming kernel argument.
+  conduit ValueT *input;
+  conduit ValueT *output;
+  // Without the annotations, kernel arguments will be inferred to be streaming
+  // kernel arguments if the kernel control interface is streaming, and
+  // vise-versa.
+  size_t n;
+  StreamingControlIP(ValueT *in_, ValueT *out_, size_t N_)
+      : input(in_), output(out_), n(N_) {}
+  // Use the 'streaming_interface' annotation on a kernel to specify it to be
+  // a kernel with streaming kernel control signals.
+  streaming_interface void operator()() const {
+    for (int i = 0; i < n; i++) {
+      output[i] = SomethingComplicated(input[i]);
+    }
+  }
+};
+
+struct CSRAgentControlIP {
+  // Use the 'register_map' annotation on a kernel argument to specify it to be
+  // a CSR Agent kernel argument.
+  register_map ValueT *input;
+  // Without the annotations, kernel arguments will be inferred to be CSR Agent
+  // kernel arguments if the kernel control interface is CSR Agent, and
+  // vise-versa.
+  ValueT *output;
+  // A kernel with CSR Agent control can also independently have streaming
+  // kernel arguments, when annotated by 'conduit'.
+  conduit size_t n;
+  CSRAgentControlIP(ValueT *in_, ValueT *out_, size_t N_)
+      : input(in_), output(out_), n(N_) {}
+  register_map_interface void operator()() const {
+    for (int i = 0; i < n; i++) {
+      output[i] = SomethingComplicated(input[i]);
+    }
+  }
+};
+
+int main(int argc, char *argv[]) {
+#if defined(FPGA_EMULATOR)
+  sycl::ext::intel::fpga_emulator_selector selector;
+#elif defined(FPGA_SIMULATOR)
+  std::string simulator_device_string =
+      "SimulatorDevice : Multi-process Simulator (aclmsim0)";
+  select_by_string selector = select_by_string{simulator_device_string};
+#else
+  sycl::ext::intel::fpga_selector selector;
+#endif
+
+  bool passed = true;
+
+  size_t count = 16;
+  if (argc > 1)
+    count = atoi(argv[1]);
+
+  if (count <= 0) {
+    std::cerr << "ERROR: 'count' must be positive" << std::endl;
+    return 1;
+  }
+
+  try {
+    // create the device queue
+    sycl::queue q(selector, dpc_common::exception_handler);
+
+    // make sure the device supports USM device allocations
```

Comment:
You need to clarify if you require USM or not

Reply:
USM is needed and desired in IPA - see above discussion about A10